components/gitpod-protocol/src/protocol.ts

```diff
@@ -863,14 +863,26 @@ export namespace PrebuiltWorkspaceContext {
     }
 }
 
+export interface WithEditorContext extends WorkspaceContext {
+    ide: string;
+    useLatest?: boolean;
+    referrer?: string;
+}
+
+export namespace WithEditorContext {
+    export function is(context: any): context is WithEditorContext {
+        return context && "ide" in context;
+    }
+}
+
 export interface WithReferrerContext extends WorkspaceContext {
     referrer: string;
     referrerIde?: string;
 }
 
 export namespace WithReferrerContext {
     export function is(context: any): context is WithReferrerContext {
-        return context && "referrer" in context;
+        return context && "referrer" in context && !("ide" in context);
```

Comment:
`&& !WithEditorContext.is(context)`?

Reply:
How about using `#referrer` and remove checkes above?

Reply:
Could you elaborate please? I did not grasp an idea, sorry.

Reply:
`#editor` works like `#referrer` we can extend `#referrer` and remove `#editor`.

Now `#referrer` only works for client (referrer) `vscode` `vscode-insiders` and `jetbrains-gateway`, and IDE option can only be one of these below 👇 . 
https://github.com/gitpod-io/gitpod/blob/330ddb2fd2d27e83e9eae400bc4ec96af91b98d2/chart/templates/server-ide-configmap.yaml#L109
We can remove this check and make referrer to be anything like `bowser-extension` `dashboard-project` `readme` etc and IDEs with any IDE. They are worth tracking too

Or extend `ide-configmap` to support `browser-extension`

Reply:
Or with `#editor` if referrer is not define, we do not send track here, and consider `#eidor` as make to choose IDE

Reply:
I think we conflate 2 issue here:
- https://github.com/gitpod-io/gitpod/issues/7525 is about allowing a user to open a workspace with given editor, we don't care about referrer
- https://github.com/gitpod-io/gitpod/issues/9434 seems to be just to enable `browser-extension`

For first issue we don't care about referrer and should not reuse it, it should be very simple `#editor:code` user is done.

For second indeed maybe we can consider to lift off some constraints. Initially we introduced to ensure that JB users via Gateway never can start a workspace with no JB product. It still should hold even we remove a constraint. You will need to check GW plugin to ensure it.

Maybe let's reopen https://github.com/gitpod-io/gitpod/issues/7525, drop `referrer` from this PR for the editor context and fix it? And then open another PR which lifts off constraints of  `referrer` prefix in a way that it does not break JB users and enables `browser-extension`?

Reply:
This PR introduces the `editor` parameter, which have three parameters, `ide`, `version`, and `referrer`, but except `ide` all of them are **optional**

So it can be used to solve #7525, just use `#ediror:code/http://xxxx`
It can also be used to solve #9434 with just `#editor:code:stable:browser-extension/http://xxxx`
It can also be used to solve any need to temporarily switch IDEs, for example, we may introduce another feature later, where we can hold down the option key in the Open workspace (select url) to temporarily select the current open IDE

Reply:
Let's solve one issue at the time instead one technical solution for everything and don't be to worry about all possible which we will need somewhen 🙏 

> It can also be used to solve https://github.com/gitpod-io/gitpod/issues/9434 with just #editor:code:stable:browser-extension/http://xxxx

Let's use `referrer` still there, but introduce a new kind `browser-extension` which allows any IDE and use `code` as default. We don't want to allow users to specify random referrers. Please also prioritise browser-extension support as minor. It is not a primary onboarding experience.

> So it can be used to solve https://github.com/gitpod-io/gitpod/issues/7525, just use #ediror:code/http://xxxx

We don't need uncontrolled referrer here at all. It also not nice syntax I would go with something like `#editor:code&referrer=.../https://`, but then it is better to use `referrer` prefix with new kinds. I suggest we go with an initial proposal here `editor:<ide>(:<version>)?`. Adding analytics is not bad then to see whether it is used, but it should not be mixed with `ide_referrer` event anymore then, something like `ide_editor_prefix`.

> It can also be used to solve any need to temporarily switch IDEs, for example, we may introduce another feature later, where we can hold down the option key in the Open workspace (select url) to temporarily select the current open IDE https://github.com/gitpod-io/gitpod/issues/9433#issuecomment-1105694971

From the dashboard, we should not rely on prefix parsers, but improve APIs to allow passing concrete IDE to create or start workspace methods. This one actually sounds like most important issue of all mentioned above. But should rather align with @loujaybee and @andreafalzetti what is goal for onboarding right now.

Reply:
> I suggest we go with an initial proposal here editor:<ide>(:<version>)?

Go with this. will create a PR to open workspace with custom IDE via API later. 

If PR above is created, and PR to add referrer `browser-extension` is created, why we need this PR? No one will use it.🙈

Feel free to close this PR @akosyakov

Reply:
> If PR above is created, and PR to add referrer browser-extension is created, why we need this PR? No one will use it.🙈

- People will add to README if a project is configured only to work nicely with VS Code for instance.
- I already would like to use it on gitpod-io/gitpod, since I usually have IntelliJ one, but would like to have a workspace only with VS Code 🙏 

I think we will need to add some docs about this prefix somewhere. @loujaybee

Reply:
Ah, sorry I misunderstand. You say you will do 2 PRs instead? Then yes just close this PR and open 2 others.